theano/tensor/basic.py

```diff
@@ -4372,7 +4372,7 @@ def infer_shape(self, node, ishapes):
                     if ele == -1:
                         requ[i] = missing
             elif crit == 1:  # we reshape to -1
-                requ = [mul(*ishapes[0])]
+                requ = [mul(*ishapes[0])] if ishapes[0] else [1]
```

Comment:
We can wait for #3420 to be fixed before merge. There is a test that Sina commented out because of this bug. We could put it back in before merge if #3420 is fixed before. Otherwise, everything seams to be done for the conversion of Flatten to Reshape.